[4.0] Add prepared statements for plg_fields_sql

HLeithner opened the pull request:
This PR tries to move the SQL custom field into a prepared statement, at least for the values part.

### Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.


### Testing Instructions

* Create a SQL custom field
* Use several SQL queries in the custom field
* check the frontend if the expected result could be found.



### Expected result
Nothing changed.

HLeithner committed:
Add prepared statements for plg_fields-sql

alikon commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
 
 // Run the query with a having condition because it supports aliases
-$db->setQuery($query . ' having value in (' . trim($condition, ',') . ')');
+$sql .= ' HAVING VALUE IN (' . implode(bindNames) . ')';
```
`$sql .= ' HAVING VALUE IN (' . implode(bindNames) . ')'`   need 1 dollar ($) before bindNames :smile

HLeithner committed:
Fix missing $.

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-fields-sql

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
-$db->setQuery($query . ' having value in (' . trim($condition, ',') . ')');
+$sql .= ' HAVING VALUE IN (' . implode($bindNames) . ')';
+
+$query->setQuery($sql);
```
```suggestion
$db->setQuery($sql);
```

SharkyKZ replied:
Move this inside `try`.

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
+
+$query->setQuery($sql);
 
 try
```
```suggestion
try
{
	$db->setQuery($query);
```

SharkyKZ replied:
Inside `catch` change `$query` to the original query (`$fieldParams->get('query', '')`).

SharkyKZ replied:
Still need to make the proposed change. Otherwise it's running an empty query.

HLeithner replied:
I changed this as you wish but this doesn't make any different because $query already got the $sql by ->setQuery()...

Anyway it's now more readable.

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
 
 // Run the query with a having condition because it supports aliases
-$db->setQuery($query . ' having value in (' . trim($condition, ',') . ')');
+$sql .= ' HAVING VALUE IN (' . implode($bindNames) . ')';
```
```suggestion
$sql .= ' HAVING ' . $db->quoteName('value') . ' IN (' . implode(',', $bindNames) . ')';
```
(value is the column alias in this case).

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
-}
-
-$query = $fieldParams->get('query', '');
+$bindNames = $query->bindArray((array) $value, ParameterType::INTEGER);
```
Don't assume integer, it could be a string too. Perhaps this needs to be configurable in the plugin?

HLeithner replied:
In this case I think it should be string even if integer. The complete field is dangerous...

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -17,24 +18,16 @@
 }
 
-$db        = Factory::getDbo();
-$value     = (array) $value;
```
Revert this. It's used as array in `in_array()` at the bottom.

HLeithner committed:
Fix catch block, implode, array and parameterType

SharkyKZ commented on plugins/fields/sql/tmpl/sql.php:
```diff
@@ -43,6 +37,7 @@
 catch (Exception $e)
 {
 	// If the query failed, we fetch all elements
+	$query->setQuery($fieldParams->get('query', ''));
```
You can remove this and just do `$db->setQuery($fieldParams->get('query', ''))` below. Otherwise this  should be a new query object without bindings.

HLeithner committed:
Finally...

HLeithner commented:
Hope it's ok now @shanginn

HLeithner committed:
Convert plugin system privacyconsent to prepared statements

SharkyKZ commented:
It's still running an empty query.

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Add missing prepared parameter

HLeithner committed:
Set the query before executing...

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-fields-sql

HLeithner committed:
Update plugins/system/privacyconsent/privacyconsent.php

Co-Authored-By: Quy <quy@fluxbb.org>

HLeithner committed:
Split core.js and remove webcomponent

HLeithner committed:
Add webcomponent loader original copy from webcomponents/polyfills

https://github.com/webcomponents/polyfills/blob/14c7b973762c22e9b8141dcffe530219a8ba3cbc/packages/webcomponentsjs/webcomponents-loader.js

HLeithner committed:
Convert webcomponent to es6 and integrate into Joomla API preserving copyright.

Based on https://github.com/joomla/joomla-cms/blob/470851b1c7020a9b5bb212dc0328d1600381653a/build/media_source/system/js/core.es6.js

HLeithner committed:
Merge branch '4.0-dev' of github.com:HLeithner/joomla-cms into 4.0-dev

HLeithner committed:
Merge remote-tracking branch 'origin/4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into 4.0-dev

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

# Conflicts:
#	plugins/system/privacyconsent/privacyconsent.php

HLeithner committed:
Revert "Convert webcomponent to es6 and integrate into Joomla API preserving copyright."

This reverts commit 109432b

HLeithner committed:
Revert "Add webcomponent loader original copy from webcomponents/polyfills"

This reverts commit 7a1e963

HLeithner committed:
Revert "Split core.js and remove webcomponent"

This reverts commit 3e40e54

HLeithner committed:
Fix reference

HLeithner committed:
cs

wilsonge committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

HLeithner committed:
Fix imports

Commit:
Merge branch '4.0-dev' into prepared-plg-system-privacyconsent

Commit:
Merge branch '4.0-dev' into prepared-plg-fields-sql

HLeithner committed:
Merge branch '4.0-dev' into prepared-plg-fields-sql

HLeithner committed:
Merge branch 'prepared-plg-system-privacyconsent' of github.com:HLeithner/joomla-cms into prepared-plg-fields-sql

HLeithner committed:
Merge branch 'prepared-plg-fields-sql' of github.com:HLeithner/joomla-cms into prepared-plg-fields-sql

HLeithner committed:
Fix setQuery exception and simplify retry

HLeithner commented:
I found another problem, if $db->setQuery($query) is not in the try catch construct and the sql query is broken (for example $value is 0 which is not possible any longer) then joomla raise an exception on setQuery because we do a Driver->prepareStatement() in this function.

This is already true for the current implementation...

Also I'm not sure how the query with the HAVING clause could fail and the original query would work.

if you have a query

```select id as value, title as text from #__content```

then the having variant will always work

```select id as value, title as text from #__content HAVING VALUE IN ('1');```

if the query fails because no value column exists it will also fail without the having clause because we expect a value column...

What do I miss?

SharkyKZ commented:
We don't know what kind of query is entered so adding anything could break it. For example if entered query ends in `;` the code already fails as it results in this:

    select id as value, title as text from #__content; HAVING VALUE IN ('1');

Another case, the query could already have a `HAVING` clause.

SharkyKZ commented:
We could avoid the first scenario by loading SQL into the query object and using the API to build having clause instead of directly manipulating SQL.

HLeithner commented:
> We could avoid the first scenario by loading SQL into the query object and using the API to build having clause instead of directly manipulating SQL.

Jquery doesn't split the query into parts.

So we should keep it this way until someone builds the ultimate query builder ;)

SharkyKZ commented:
Hm, you're right. Although we could still use query methods for HAVING clause. But it's fine either way. Just lowercase and quote `value` please because it's a column name.

HLeithner committed:
Quote column value

SharkyKZ commented:
I have tested this item :white_check_mark: successfully on a952aed30b21dbe9e3c5ea531508fcd2b48791ae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25056">issues.joomla.org/tracker/joomla-cms/25056</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on a952aed30b21dbe9e3c5ea531508fcd2b48791ae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25056">issues.joomla.org/tracker/joomla-cms/25056</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25056">issues.joomla.org/tracker/joomla-cms/25056</a>.</sub>

infograf768 committed:
Merge branch '4.0-dev' into prepared-plg-fields-sql
